dm/worker/worker_test.go

```diff
@@ -170,3 +174,223 @@ func (t *testServer) TestTaskAutoResume(c *C) {
+}
+
+func (t *testWorkerEtcdCompact) TestWatchRelayStageEtcdCompact(c *C) {
+	var (
+		masterAddr   = "127.0.0.1:8291"
+		keepAliveTTL = int64(1)
+		startRev     = int64(1)
+	)
+	etcdDir := c.MkDir()
+	ETCD, err := createMockETCD(etcdDir, "host://"+masterAddr)
+	c.Assert(err, IsNil)
+	defer ETCD.Close()
+	cfg := NewConfig()
+	c.Assert(cfg.Parse([]string{"-config=./dm-worker.toml"}), IsNil)
+	cfg.KeepAliveTTL = keepAliveTTL
+
+	subtaskCfg := config.SubTaskConfig{}
+	err = subtaskCfg.DecodeFile(subtaskSampleFile, true)
+	c.Assert(err, IsNil)
+	subtaskCfg.MydumperPath = mydumperPath
+
+	etcdCli, err := clientv3.New(clientv3.Config{
+		Endpoints:            GetJoinURLs(cfg.Join),
+		DialTimeout:          dialTimeout,
+		DialKeepAliveTime:    keepaliveTime,
+		DialKeepAliveTimeout: keepaliveTimeout,
+	})
+	c.Assert(err, IsNil)
+	sourceCfg := loadSourceConfigWithoutPassword(c)
+	sourceCfg.EnableRelay = true
+	sourceCfg.RelayDir = c.MkDir()
+	sourceCfg.MetaDir = c.MkDir()
+
+	// step 1: start worker
+	w, err := NewWorker(&sourceCfg, etcdCli)
+	c.Assert(err, IsNil)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	defer w.Close()
+	go func() {
+		w.Start(true)
+	}()
+	c.Assert(utils.WaitSomething(50, 100*time.Millisecond, func() bool {
+		return w.closed.Get() == closedFalse
+	}), IsTrue)
+	// step 2: Put a relay stage to this source, then delete it
+	// put mysql config into relative etcd key adapter to trigger operation event
+	c.Assert(w.relayHolder, NotNil)
+	_, err = ha.PutSourceCfg(etcdCli, sourceCfg)
+	c.Assert(err, IsNil)
+	_, err = ha.PutRelayStageSourceBound(etcdCli, ha.NewRelayStage(pb.Stage_Running, sourceCfg.SourceID),
+		ha.NewSourceBound(sourceCfg.SourceID, cfg.Name))
+	c.Assert(err, IsNil)
+	rev, err := ha.DeleteSourceCfgRelayStageSourceBound(etcdCli, sourceCfg.SourceID, cfg.Name)
+	c.Assert(err, IsNil)
+	// check relay stage, should be running
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		return w.relayHolder.Stage() == pb.Stage_Running
+	}), IsTrue)
+	// step 3: trigger etcd compaction and check whether we can receive it through watcher
+	_, err = etcdCli.Compact(ctx, rev)
+	c.Assert(err, IsNil)
+	relayStageCh := make(chan ha.Stage, 10)
+	relayErrCh := make(chan error, 10)
+	ha.WatchRelayStage(ctx, etcdCli, cfg.Name, startRev, relayStageCh, relayErrCh)
+	select {
+	case err := <-relayErrCh:
+		c.Assert(err, Equals, etcdErrCompacted)
+	case <-time.After(300 * time.Millisecond):
+		c.Fatal("fail to get etcd error compacted")
+	}
+	// step 4: watch relay stage from startRev
+	var wg sync.WaitGroup
+	ctx1, cancel1 := context.WithCancel(ctx)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		c.Assert(w.observeRelayStage(ctx1, etcdCli, startRev), IsNil)
+	}()
+	// step 5: should stop the running relay
+	time.Sleep(time.Second)
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		return w.relayHolder.Stage() == pb.Stage_Stopped
+	}), IsTrue)
+	cancel1()
+	wg.Wait()
```

Comment:
Do we need to check whether it can run again after a new stage and cfg putted?

Reply:
If a new config is putted the worker will be closed and restart another worker, so I think we can just check whether it can quit successfully.

Reply:
OK